forest_benchmarking/quantum_volume.py

```diff
@@ -0,0 +1,262 @@
+    can adequately implement random model circuits on depth-many qubits such that the given
+    depth is 'achieved'. A model circuit depth is achieved if the sample distribution for a
+    sample of num_circuits many randomly generated model circuits of the given depth sufficiently
+    matches the ideal distribution of that circuit (See Eq. 6  of [QVol]). The frequency of
+    sampling 'heavy-outputs' is used as a measure of closeness of the circuit distributions. This
+    estimated frequency (across all sampled circuits) is reported for each depth along with a
+    bool which indicates whether that depth was achieved. The logarithm of the quantum volume is by
+    definition the largest achievable depth of the circuit; see
+    extract_quantum_volume_from_results for obtaining the quantum volume from the results
+    returned by this method.
+
+            [QVol] Validating quantum computers using randomized model circuits
+            Cross et al., arXiv:1811.12926v1, Nov 2018
+            https://arxiv.org/pdf/1811.12926.pdf
+
+    :param qc: the quantum resource whose volume you wish to measure
+    :param qubits: available qubits on which to act during measurement. Default all qubits in qc.
+    :param program_generator: a method which
+        1) takes in a quantum computer, the qubits on that
+            computer available for use, an array of sequences representing the qubit permutations
+            in a model circuit, an array of matrices representing the 2q gates in the model circuit
+        2) outputs a native quil program that implements the circuit and measures the appropriate
+            qubits in the order implicitly dictated by the model circuit representation created in
+            sample_rand_circuits_for_heavy_out.
+        The default option simply picks the smallest qubit labels and lets the compiler do the rest.
+    :param num_circuits: number of unique random circuits that will be sampled.
+    :param num_shots: number of shots for each circuit sampled.
+    :param depths: the circuit depths to scan over. Defaults to all depths from 2 to len(qubits)
+    :param achievable_threshold: threshold at which a depth is considered to be achieved.
+        Eq. 6 of [QVol] defines this to be the default of 2/3
+    :param stop_when_fail: if true, the measurement will stop after the first un-achievable depth
+    :param show_progress_bar: displays a progress bar for each depth if true.
+    :return: a list of all tuples (depth, prob_sample_heavy, is_achievable) indicating the
+        estimated probability of sampling a heavy output at each depth and whether that depth
+        qualifies as being achievable.
+    """
+
+    if num_circuits < 100:
+        warnings.warn("The number of random circuits ran ought to be greater than 100 for results "
+                      "to be valid.")
+    if qubits is None:
+        qubits = qc.qubits()
+    qubits = sorted(qubits)
```

Comment:
ok i'm back. I think you should be able to trivially remove the sorting, although it really doesn't matter since we apply permutations anyways. Not blocking merge of this pr